resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodFactory.scala

```diff
@@ -0,0 +1,229 @@
+    .getOrElse(math.max((MEMORY_OVERHEAD_FACTOR * executorMemoryMiB).toInt,
+      MEMORY_OVERHEAD_MIN_MIB))
+  private val executorMemoryWithOverhead = executorMemoryMiB + memoryOverheadMiB
+
+  private val executorCores = sparkConf.getDouble("spark.executor.cores", 1d)
+  private val executorLimitCores = sparkConf.getOption(KUBERNETES_EXECUTOR_LIMIT_CORES.key)
+
+  override def createExecutorPod(
+      executorId: String,
+      applicationId: String,
+      driverUrl: String,
+      executorEnvs: Seq[(String, String)],
+      driverPod: Pod,
+      nodeToLocalTaskCount: Map[String, Int]): Pod = {
+    val name = s"$executorPodNamePrefix-exec-$executorId"
+
+    // hostname must be no longer than 63 characters, so take the last 63 characters of the pod
+    // name as the hostname.  This preserves uniqueness since the end of name contains
+    // executorId and applicationId
+    val hostname = name.substring(Math.max(0, name.length - 63))
+    val resolvedExecutorLabels = Map(
+      SPARK_EXECUTOR_ID_LABEL -> executorId,
+      SPARK_APP_ID_LABEL -> applicationId,
+      SPARK_ROLE_LABEL -> SPARK_POD_EXECUTOR_ROLE) ++
+      executorLabels
+    val executorMemoryQuantity = new QuantityBuilder(false)
+      .withAmount(s"${executorMemoryMiB}Mi")
+      .build()
+    val executorMemoryLimitQuantity = new QuantityBuilder(false)
+      .withAmount(s"${executorMemoryWithOverhead}Mi")
+      .build()
+    val executorCpuQuantity = new QuantityBuilder(false)
+      .withAmount(executorCores.toString)
+      .build()
+    val executorExtraClasspathEnv = executorExtraClasspath.map { cp =>
+      new EnvVarBuilder()
+        .withName(ENV_EXECUTOR_EXTRA_CLASSPATH)
+        .withValue(cp)
+        .build()
+    }
+    val executorExtraJavaOptionsEnv = sparkConf
+      .get(org.apache.spark.internal.config.EXECUTOR_JAVA_OPTIONS)
+      .map { opts =>
+        val delimitedOpts = Utils.splitCommandString(opts)
+        delimitedOpts.zipWithIndex.map {
+          case (opt, index) =>
+            new EnvVarBuilder().withName(s"$ENV_JAVA_OPT_PREFIX$index").withValue(opt).build()
+        }
+      }.getOrElse(Seq.empty[EnvVar])
+    val executorEnv = (Seq(
+      (ENV_EXECUTOR_PORT, executorPort.toString),
+      (ENV_DRIVER_URL, driverUrl),
+      // Executor backend expects integral value for executor cores, so round it up to an int.
+      (ENV_EXECUTOR_CORES, math.ceil(executorCores).toInt.toString),
+      (ENV_EXECUTOR_MEMORY, executorMemoryString),
+      (ENV_APPLICATION_ID, applicationId),
+      (ENV_EXECUTOR_ID, executorId),
+      (ENV_MOUNTED_CLASSPATH, s"$executorJarsDownloadDir/*")) ++ executorEnvs)
+      .map(env => new EnvVarBuilder()
+        .withName(env._1)
+        .withValue(env._2)
+        .build()
+      ) ++ Seq(
+      new EnvVarBuilder()
+        .withName(ENV_EXECUTOR_POD_IP)
+        .withValueFrom(new EnvVarSourceBuilder()
+          .withNewFieldRef("v1", "status.podIP")
+          .build())
+        .build()
```

Comment:
There is no stable hostname for a Kubernetes Pod unless we put the Pod behind a Service. We can't put executors behind services because Services are considered as heavyweight objects, and we want to constrain the number of them we have to make.

Reply:
Hm, the above comment should be more precise.

DNS is handled by kube-dns in any given Kubernetes context. (kube-dns is an optional component, but in future commits it will become clear that Spark will require kube-dns to be installed on a cluster, and we will document as such). kube-dns creates a DNS entry for services that route to the IPs of pods in the system. But kube-dns does not create a DNS entry for every pod, unless there is a service that maps to it.

Reply:
There are various parts of spark which assert that a hostname is used and not IP - that is, spark makes the assumption about executors and drivers being reachable via hostnames.
Am I right in understanding that this assumption does not hold for kubernetes case ?
That is, containers need not have a routable hostname assigned to them ?

Reply:
Just to clarify, the hostname need not be stable - but needs to be stable only for the duration the container is 'alive' : and is not immediately reused (in case some other executor container comes up with exact same hostname after previous exits).

Reply:
Containers in Kubernetes will not have resolvable hostnames unless we specifically add a DNS entry for it via a service, which is heavyweight and should be done as infrequently as possible. We will need to handle the cases where this breaks Spark as they come up.

I'm curious as to what some example cases are where the hostname matters - we have tackled a few of them on our fork.

Reply:
Please check for references to `Utils.checkHost` - which requires hostnames.
The enforcement was kept slightly liberal on purpose, but spark does not support IP's.

Reply:
Did a quick scan - looks like we use that call in many places. But given our heavy reliance on hostnames, how would this manifest in practice? Is it the case where Spark shouldn't be working at all? We haven't seen any problems specifically with the hostnames of executors, and we recently fixed needing a hostname for the driver by putting the driver behind a service.

Reply:
I forgot what the exact reasons were when I introduced this (way back in 0.6 iirc).
In particular ipv4 -> ipv6 was one of the culprits iirc (which is why we check for ':' in hostname, but not enforce avoidance of dotted decimal pattern) which was messing up host:port parsing.
IIRC there were other corner cases which caused issues too - particularly with locality parsing, multi-honed machines, etc.

Reply:
Kubernetes networking is a pretty interesting topic in and of itself, and I foresee part of the challenge of the project will be dealing with these unexpected and unpredictable situations where Spark's assumptions about networking won't hold in the Kubernetes setting. My two cents is that we can only make headway on this project by using what we know to be true about Kubernetes networking - the fact that we can't use hostnames for all executors and have to communicate over IP addresses - and to then tackle networking bugs as they come up. On both production clusters and our integration test environment, we have caught some of these issues as they have come up and patched them accordingly already, and we can continue to do so.

@mridulm - is this a reasonable way forward, or how would you suggest we proceed otherwise?

@foxish do you have any thoughts on this discussion thread?

Reply:
FWIW we've been running the combination of Apache master branch with https://github.com/apache-spark-on-k8s/spark (both synced ~weekly) on internal clusters and have had only one problem with hostname requirements:

when SPARK-21642 merged it broke the way executors were registering with the driver which we fixed by giving the driver RPC port a hostname: https://github.com/apache-spark-on-k8s/spark/pull/483

Other than that, we haven't seen any issues with executors listening on and being identified by IP addresses instead of hostnames.

Reply:
@mccheah @ash211 If it is not possible to associate hostnames with IP, then we will need to find a way to work around the restriction.
The hostname requirement was a practical solution to an existing limitations of that time, nothing inherently limiting to spark core.

We could proceed by clearly documenting the constraints of the integration, and possibly relax it future as we iterate and improve.

If we are to support only IP's, things which come to mind are :
* No support for IPv6
* No support for multi-honed machines.
* Distributed stores (where relevant) should always return IP's for locality information.
* Rack awareness config should work with IP's.

Possibly add more as we get to it, and document this clearly for our users.
How does that sound ?

Reply:
To add, if the list above is not accurate (needs additions/mods/deletions), that is fine - I was listing potential concerns from top of my head.

Reply:
@mridulm  In response to each of your four points:

- IPv6: IP addresses are strictly managed by the Kubernetes framework, so it's unlikely we're going to run into differences between Ipv4 and Ipv6 in different Kubernetes clusters. We should assume that one of these two address types are being used across all clusters and work with that. So far I've only seen Kubernetes assign IPv4 addresses.
- No support for multihomed machines of multi-routable IP's: Again, since Kubernetes is managing the IP address and routability of pods, we can understand what the framework will do and work with that. I don't think the framework does anything fancy in this space, but maybe @foxish or others have ideas?
- Distributed stores: We've thought about this and have some work done on our fork in regards to this that we will eventually contribute back here. @kimoonkim has done some work on this. The short version of the situation we've had to work around is that the container that runs the Spark processes now has a different IP address from the HDFS node that might be colocated with it physically.
- Rack awareness - @kimoonkim has also done work on this, and similar concerns to the point above have come up.

As a general note, locality is non-trivial in Kubernetes because no two pods will ever share the same IP address, and pods do not share the same IP address as the physical host that is running it. The Kubernetes code needs to be intelligent about knowing which pods are co-located on the same underlying Kubelet. And of course, it's reasonable to believe that the above four considerations are not exhaustive, but we'll address unforeseen factors as they come up.

Reply:
Thanks for clarifying @mccheah.

One clarification though:
> IP addresses are strictly managed by the Kubernetes framework, so it's unlikely we're going to run into differences between Ipv4 and Ipv6 in different Kubernetes clusters. We should assume that one of these two address types are being used across all clusters and work with that

When I referred to IPv6 support, I was not referring to interop between IPv4 and IPv6 (that is a different can of worms !). What I wanted to clarify was that IPv6 is not supported as IP (supported via hostnames though).
If all we are getting are IPv4 ip's, I do not see any obvious issues.

> The Kubernetes code needs to be intelligent about knowing which pods are co-located on the same underlying Kubelet

This would be critical to get HOST locality level working properly. The performance implications of not getting this right would be non trivial.

Looking forward to how this has been addressed !

Reply:
Was doing another pass over this thread. I think Matt covered this mostly. IPv6 support is being added in https://github.com/kubernetes/features/issues/508 but will be in alpha for now. We should target testing with that once that goes beta (probably Kubernetes 1.10 timeframe). As for routing, that's entirely handled by k8s network plugins and we shouldn't have to do anything special to take advantage of it.

Conceptually, the notion of needing a network identity like DNS is separated from containers/pods themselves by design. We *could* if necessary, create [headless services](https://kubernetes.io/docs/concepts/services-networking/service/#headless-services) if there's a strong need for a DNS name for each executor, but that would be a heavyweight approach that I would prefer we didn't unless we absolutely had to.  

Lastly, +1 on the locality comment. I do think that's very important and there's ways we can use the k8s control plane to figure out a global view - different executors and their corresponding physical (or virtual nodes).